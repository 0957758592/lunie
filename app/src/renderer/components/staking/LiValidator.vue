<template>
  <tr class="data-table__row li-validator">
    <td class="data-table__row__info">
      <img
        v-if="validator.keybase"
        :src="validator.keybase.avatarUrl"
        class="data-table__row__info__image"
        width="48"
        height="48"
      >
      <img
        v-else
        class="
          data-table__row__info__image
          data-table__row__info__image--no-img
        "
        src="~assets/images/validator-icon.svg"
        width="48"
        height="48"
      >
      <div class="data-table__row__info__container">
        <span
          v-tooltip.top="status"
          :class="statusColor"
          class="data-table__row__info__container__status"
        />
        <router-link
          :to="{
            name: 'validator',
            params: { validator: validator.operator_address }
          }"
          :class="styles"
          class="data-table__row__info__container__name"
        >
          {{ validator.description.moniker }}
        </router-link>
        <div class="data-table__row__info__container__description">
          <short-bech32 :address="validator.operator_address" />
        </div>
      </div>
    </td>
    <td class="li-validator__delegated-steak">
      {{ yourVotes.isGreaterThan(0) ? num.shortNumber(yourVotes) : `--` }}
    </td>
    <td class="li-validator__voting-power">
      {{
        validator.percent_of_vote
          ? num.percent(validator.percent_of_vote)
          : `--`
      }}
    </td>
    <td class="li-validator__commission">
      {{ validator.commission ? num.percent(validator.commission) : `--` }}
    </td>
    <td class="li-validator__uptime">
      {{ validator.uptime ? num.percent(validator.uptime) : `--` }}
    </td>
  </tr>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { calculateTokens, ratToBigNumber } from "scripts/common"
import ShortBech32 from "common/ShortBech32"
import BigNumber from "bignumber.js"
export default {
  name: `li-validator`,
  components: {
    ShortBech32
  },
  props: {
    validator: {
      type: Object,
      required: true
    },
    disabled: {
      type: Boolean,
      required: true
    }
  },
  data: () => ({ num }),
  computed: {
    ...mapGetters([
      `delegates`,
      `committedDelegations`,
      `distribution`,
      `bondDenom`,
      `session`,
      `lastHeader`
    ]),
    yourVotes() {
      return this.committedDelegations[this.validator.operator_address]
        ? BigNumber(
          num
            .atoms(
              calculateTokens(
                this.validator,
                this.committedDelegations[this.validator.operator_address]
              )
            )
            .toString()
        )
        : BigNumber(0)
    },
    styles() {
      let value = ``
      if (this.validator.isValidator) value += `li-validator-validator `
      return value
    },
    delegateType() {
      return this.validator.revoked
        ? `Revoked`
        : this.validator.isValidator
          ? `Validator`
          : `Candidate`
    },
    powerRatio() {
      return ratToBigNumber(this.validator.tokens)
        .div(this.delegates.globalPower)
        .toNumber()
    },
    status() {
      // status: jailed
      if (this.validator.revoked)
        return `This validator has been jailed and is not currently validating`

      // status: inactive
      if (parseFloat(this.validator.voting_power) === 0)
        return `This validator does not have enough voting power yet and is inactive`

      // status: active
      return `This validator is actively validating`
    },
    statusColor() {
      // status: jailed
      if (this.validator.revoked) return `red`

      // status: inactive
      if (parseFloat(this.validator.voting_power) === 0) return `yellow`

      // status: active
      return `green`
    },
    rewards() {
      if (!this.session.signedIn) return null
      const validatorRewards = this.distribution.rewards[
        this.validator.operator_address
      ]
      return validatorRewards
        ? num.shortNumber(num.atoms(validatorRewards[this.bondDenom]) || 0)
        : null
    }
  },
  watch: {
    lastHeader: {
      immediate: true,
      handler() {
        if (this.yourVotes > 0) {
          this.$store.dispatch(
            `getRewardsFromValidator`,
            this.validator.operator_address
          )
        }
      }
    }
  }
}
</script>
