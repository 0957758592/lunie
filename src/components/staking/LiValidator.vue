<template>
  <tr
    class="li-validator"
    :data-moniker="validator.description.moniker"
    @click="
      $router.push({
        name: 'validator',
        params: { validator: validator.operator_address }
      })
    "
  >
    <td class="data-table__row__info">
      <img
        v-if="validator.keybase && validator.keybase.avatarUrl"
        :src="validator.keybase.avatarUrl"
        class="li-validator-image"
        :alt="`validator logo for ` + validator.description.moniker"
      />
      <img
        v-else
        class="li-validator-image"
        src="~assets/images/validator-icon.svg"
        alt="generic validator logo - graphic triangle supporting atom token"
      />
      <div class="validator-info">
        <h3 class="li-validator-name">{{ validator.description.moniker }}</h3>
        <div v-if="validator.my_delegations > 0">
          <h4>
            {{
              validator.my_delegations
                ? num.shortDecimals(num.atoms(validator.my_delegations))
                : null
            }}
          </h4>
          <h5 v-if="validator.rewards > 0">
            {{
              validator.rewards
                ? `+` + num.shortDecimals(num.atoms(validator.rewards))
                : `--`
            }}
          </h5>
        </div>
      </div>
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'expectedReturns' }">
      {{
        validator.expectedReturns
          ? num.percent(validator.expectedReturns)
          : `--`
      }}
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'voting-power' }">
      {{ validator.tokens ? percentOfVotingPower : `--` }}
    </td>
  </tr>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import BN from "bignumber.js"
export default {
  name: `li-validator`,
  props: {
    validator: {
      type: Object,
      required: true
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({ num }),
  computed: {
    ...mapGetters([
      `delegates`,
      `distribution`,
      `session`,
      `lastHeader`,
      `pool`
    ]),
    percentOfVotingPower() {
      return num.percent(
        BN(this.validator.tokens)
          .div(this.pool.pool.bonded_tokens)
          .toFixed(4)
      )
    }
  }
}
</script>
<style scoped>
.li-validator {
  padding: 0.5rem 1rem;
  margin-bottom: 0.25rem;
  border-bottom: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
}

.li-validator:last-child {
  border-bottom: none;
}

.validator-info {
  display: flex;
  flex-direction: column;
  padding-left: 1rem;
  text-overflow: ellipsis;
}

.li-validator h4,
.li-validator h5 {
  font-size: var(--sm);
  display: inline-block;
}

.li-validator h5 {
  color: var(--success);
}

.li-validator:hover {
  cursor: pointer;
  background: var(--hover-bg);
  color: var(--bright);
}

.li-validator-name {
  font-size: 1rem;
  line-height: 18px;
  font-weight: 500;
  white-space: pre-wrap;
  color: var(--bright);
  display: inline-block;
}

.li-validator-image {
  border-radius: 0.25rem;
  height: 2.5rem;
  min-width: 2.5rem;
  border: 1px solid var(--bc-dim);
}
</style>
