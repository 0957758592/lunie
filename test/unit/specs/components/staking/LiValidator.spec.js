import { shallowMount, createLocalVue } from "@vue/test-utils"
import LiValidator from "src/components/staking/LiValidator"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => {})

describe(`LiValidator`, () => {
  let wrapper, $store

  const validator = {
    operator_address: `cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au`,
    pub_key: `cosmosvalpub1234`,
    jailed: false,
    tokens: `14`,
    delegator_shares: `14`,
    description: {
      website: `www.monty.ca`,
      details: `Mr Mounty`,
      moniker: `mr_mounty`,
      country: `Canada`
    },
    status: 2,
    bond_height: `0`,
    bond_intra_tx_counter: 6,
    proposer_reward_pool: null,
    commission: `0.05`,
    prev_bonded_shares: `0`,
    signing_info: {
      start_height: 0,
      index_offset: 465400,
      jailed_until: `1970-01-01T00:00:00Z`,
      missed_blocks_counter: 122
    },
    uptime: 0.98778883,
    expectedReturns: 0.13
  }

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        delegates: { delegates: [] },
        committedDelegations: {},
        distribution: {
          rewards: {}
        },
        session: {
          signedIn: true
        },
        bondDenom: `stake`,
        lastHeader: ``,
        pool: {
          pool: {
            bonded_tokens: 1000
          }
        }
      }
    }

    wrapper = shallowMount(LiValidator, {
      localVue,
      propsData: {
        validator,
        disabled: false
      },
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should show the voting power`, () => {
    expect(wrapper.html()).toContain(`1.40%`)
  })

  it(`works if user is not signed in`, () => {
    $store.getters.session.signedIn = false

    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
