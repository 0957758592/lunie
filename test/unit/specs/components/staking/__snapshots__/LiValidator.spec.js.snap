// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiValidator has the expected html structure 1`] = `
<tr
  class="data-table__row li-validator"
  data-moniker="mr_mounty"
>
  <td
    class="data-table__row__info"
  >
    <img
      class="
        data-table__row__info__image
        data-table__row__info__image--no-img
      "
      src="~assets/images/validator-icon.svg"
    />
     
    <div
      class="data-table__row__info__container"
    >
      <span
        class="data-table__row__info__container__status green"
      />
       
      <span
        class="data-table__row__info__container__name"
      >
        
        mr_mounty
      
      </span>
    </div>
  </td>
   
  <td
    class="li-validator__delegated-steak"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__rewards"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__voting-power"
  >
    
    1.40%
  
  </td>
   
  <td
    class="li-validator__commission"
  >
    
    5.00%
  
  </td>
   
  <td
    class="li-validator__uptime"
  >
    
    98.78%
  
  </td>
   
  <td
    class="li-validator__yield"
  >
    
    13.00%
  
  </td>
</tr>
`;

exports[`LiValidator should display uptime as double-dash 1`] = `
<tr
  class="data-table__row li-validator"
  data-moniker="mr_mounty"
>
  <td
    class="data-table__row__info"
  >
    <img
      class="
        data-table__row__info__image
        data-table__row__info__image--no-img
      "
      src="~assets/images/validator-icon.svg"
    />
     
    <div
      class="data-table__row__info__container"
    >
      <span
        class="data-table__row__info__container__status green"
      />
       
      <span
        class="data-table__row__info__container__name"
      >
        
        mr_mounty
      
      </span>
    </div>
  </td>
   
  <td
    class="li-validator__delegated-steak"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__rewards"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__voting-power"
  >
    
    1.40%
  
  </td>
   
  <td
    class="li-validator__commission"
  >
    
    5.00%
  
  </td>
   
  <td
    class="li-validator__uptime"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__yield"
  >
    
    13.00%
  
  </td>
</tr>
`;

exports[`LiValidator works if user is not signed in 1`] = `
<tr
  class="data-table__row li-validator"
  data-moniker="mr_mounty"
>
  <td
    class="data-table__row__info"
  >
    <img
      class="
        data-table__row__info__image
        data-table__row__info__image--no-img
      "
      src="~assets/images/validator-icon.svg"
    />
     
    <div
      class="data-table__row__info__container"
    >
      <span
        class="data-table__row__info__container__status green"
      />
       
      <span
        class="data-table__row__info__container__name"
      >
        
        mr_mounty
      
      </span>
    </div>
  </td>
   
  <td
    class="li-validator__delegated-steak"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__rewards"
  >
    
    --
  
  </td>
   
  <td
    class="li-validator__voting-power"
  >
    
    1.40%
  
  </td>
   
  <td
    class="li-validator__commission"
  >
    
    5.00%
  
  </td>
   
  <td
    class="li-validator__uptime"
  >
    
    98.78%
  
  </td>
   
  <td
    class="li-validator__yield"
  >
    
    13.00%
  
  </td>
</tr>
`;
